ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchive.java

```diff
@@ -160,28 +160,37 @@ public Optional<MutableWorldState> getMutableSnapshot(final Hash blockHash) {
   @Override
   public Optional<MutableWorldState> getMutable(
       final Hash rootHash, final Hash blockHash, final boolean shouldPersistState) {
+    final Optional<MutableWorldState> mutableWorldstate;
     if (shouldPersistState) {
-      return getMutable(rootHash, blockHash);
+      mutableWorldstate = getMutable(rootHash, blockHash);
     } else {
-      return trieLogManager
-          .getBonsaiCachedWorldState(blockHash)
-          .or(
-              () ->
-                  blockchain
-                      .getBlockHeader(blockHash)
-                      .filter(
-                          header -> {
-                            if (blockchain.getChainHeadHeader().getNumber() - header.getNumber()
-                                >= trieLogManager.getMaxLayersToLoad()) {
-                              LOG.warn(
-                                  "Exceeded the limit of back layers that can be loaded ({})",
-                                  trieLogManager.getMaxLayersToLoad());
-                              return false;
-                            }
-                            return true;
-                          })
-                      .flatMap(header -> snapshotOrLayeredWorldState(blockHash, header)));
+      mutableWorldstate =
+          trieLogManager
+              .getBonsaiCachedWorldState(blockHash)
+              .or(
+                  () ->
+                      blockchain
+                          .getBlockHeader(blockHash)
+                          .filter(
+                              header -> {
+                                if (blockchain.getChainHeadHeader().getNumber() - header.getNumber()
+                                    >= trieLogManager.getMaxLayersToLoad()) {
+                                  LOG.warn(
+                                      "Exceeded the limit of back layers that can be loaded ({})",
+                                      trieLogManager.getMaxLayersToLoad());
+                                  return false;
+                                }
+                                return true;
+                              })
+                          .flatMap(header -> snapshotOrLayeredWorldState(blockHash, header)));
     }
+    return mutableWorldstate.map(
```

Comment:
👍

Reply:
I will do as before. I cannot see easy way to manage that . I wanted to create an OptionalMutableWorldstate with only a ifPresent that close the ressource automatically but seems to be a big and complex modification

I will also reduce the scope to the transaction pool and we will see another PR for the blockReplay part